pkg/proxy/benchmarking/benchmark.sh

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env bash
+
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+set -o errexit
+set -o nounset
+set -o pipefail
+# shellcheck disable=SC1073
+# shellcheck disable=SC1009
+
+FILE=${0}
+RED="\033[91m"
+BLUE="\033[94m"
+ENDCOLOR="\033[0m"
+
+function help {
+    printf "\n"
+    printf "This script is expected to run twice for comparing initial and final benchmarks."
+    printf "\n\tFirst run for establishing a baseline benchmark this should be done by pointing"
+    printf "\n\tthe repository to master/main branch of the upstream."
+    printf "\n\tSecond run for generating benchmarks after code optimization and finally comparing "
+    printf "\n\tthe target with baseline benchmark."
+    printf "\n\nUsage: %s [-b] [-c count] [-f function] [-h]\n" "${FILE}"
+    printf "\t-b generate baseline benchmark, generates target benchmark if '-b' not specified.\n"
+    printf "\t-c count, run each benchmark n times.\n"
+    printf "\t-f function, regular expression for the function to run benchmark on.\n"
+    printf "\t-h help.\n"
+    printf "\nExample:\n\t %s -b -c 25 -f BenchmarkWriter\n" "${FILE}"
+    exit 1
+}
+
+function info_message(){
+    echo -e "[ ${BLUE}INFO${ENDCOLOR} ] ${1}"
+}
+
+function error_message(){
+    echo -e "[ ${RED}ERROR${ENDCOLOR} ] ${1}"
+}
+
+function benchmark() {
+    mode=$1
+    count=$2
+    function=$3
+
+    # shellcheck disable=SC2027
+    # shellcheck disable=SC2140
+    info_message "mode: ${mode}"
+
+    # shellcheck disable=SC2140
+    info_message "running: go test --bench=""${function}"" --benchmem --count=""${count}"" --run=^# > ""${mode}".report""
+
+    # run the benchmark
+    GOMAXPROCS=1 go test --bench="${function}" --benchmem --count="${count}" --run=^# > "${mode}.report"
+
+    # generate comparison if the mode is target
+    if [[ "${mode}" == "target" ]]; then
+        # exit if baseline benchmark doesn't exist
+        if [[ ! -f "baseline.report" ]]; then
+            # shellcheck disable=SC1068
+            error_message "no baseline benchmark to compare with, run '${FILE} -b' to generate baseline benchmark for comparison\n"
+            exit 1
+        fi
+
+        # generate comparison
+        info_message "running benchstat: benchstat baseline.report target.report"
+        benchstat baseline.report target.report
+    fi
+}
+
+mode="target"
+count=100
+function="."
+
+while getopts hbc:f: flag
+do
+    # shellcheck disable=SC2220
```

Comment:
We should note why it is valid to exclude these. I'm not convinced it is in this case.

Reply:
Same for above